README.md
django-guac-auth
=======================

Django app to facilitate setting up Guacamole database items

[![Build Status](https://travis-ci.org/nimbis/django-guac-auth.svg?branch=master)](https://travis-ci.org/nimbis/django-guac-auth)

[![Coverage](https://coveralls.io/repos/nimbis/django-guac-auth/badge.png?branch=master)](https://coveralls.io/r/nimbis/django-guac-auth?branch=master)


Requirements
------------

* django >= 1.8

Installation
------------

* Run `pip install django-guac-auth` or download this package and run `python setup.py install`

* Ensure that `guac_auth` is in your INSTALLED APPS

Contributing
------------

See the [Contributing Guidelines](CONTRIBUTING.md).

History
-------
See file CHANGES.

Usage
-----

Several utility methods are provided.

    # Set up an RDP connection, guac user, and needed permissions
    quick_rdp(guac_username="guactestuser", guac_password="guactest",
              username="Administrator", password="somepass",
              hostname="somehost")
    
    # Set up an RDP connection with the provided parameters
    quick_rdp_conn(username="Adminstrator", password="somepass",
                   hostname="somehost"):
    
    # Set up a Guacamole user
    quick_guac_user(username="guactestuser", password="guactest"):
    
    # Destroy all bits set up by quick_rdp()
    quick_rdp_destroy(guac_username="guactestuser", username="Administrator",
                      hostname="somehost", cleanup_user=True,
                      cleanup_connection=True):


CHANGES
1.0.2 (2017-01-13):

    * Added support for Django 1.9 and 1.10.

1.0.1 (2016-07-18):

    * First actual release, (I don't think 1.0.0 was ever
      released---at least it wasn't tagged).

v1.0.0:

    * Initial version.


tests/settings.py
DEBUG = True

TEST_RUNNER = 'django.test.runner.DiscoverRunner'


ADMINS = (
    # ('Your Name', 'your_email@example.com'),
)

MANAGERS = ADMINS

DATABASES = {
    'default': {
        'ENGINE': 'django.db.backends.sqlite3',
        'NAME': 'db.sqlite3',
        'TEST': {
            'NAME': 'db-test.sqlite3',
        },
    },
}

# Hosts/domain names that are valid for this site; required if DEBUG is False
# See https://docs.djangoproject.com/en/1.5/ref/settings/#allowed-hosts
ALLOWED_HOSTS = []

# Local time zone for this installation. Choices can be found here:
# http://en.wikipedia.org/wiki/List_of_tz_zones_by_name
# although not all choices may be available on all operating systems.
# In a Windows environment this must be set to your system time zone.
TIME_ZONE = 'America/Chicago'

# Language code for this installation. All choices can be found here:
# http://www.i18nguy.com/unicode/language-identifiers.html
LANGUAGE_CODE = 'en-us'

SITE_ID = 1

# If you set this to False, Django will make some optimizations so as not
# to load the internationalization machinery.
USE_I18N = True

# If you set this to False, Django will not format dates, numbers and
# calendars according to the current locale.
USE_L10N = True

# If you set this to False, Django will not use timezone-aware datetimes.
USE_TZ = True

# Absolute filesystem path to the directory that will hold user-uploaded files.
# Example: "/var/www/example.com/media/"
MEDIA_ROOT = ''

# URL that handles the media served from MEDIA_ROOT. Make sure to use a
# trailing slash.
# Examples: "http://example.com/media/", "http://media.example.com/"
MEDIA_URL = ''

# Absolute path to the directory static files should be collected to.
# Don't put anything in this directory yourself; store your static files
# in apps' "static/" subdirectories and in STATICFILES_DIRS.
# Example: "/var/www/example.com/static/"
STATIC_ROOT = ''

# URL prefix for static files.
# Example: "http://example.com/static/", "http://static.example.com/"
STATIC_URL = '/static/'

# Additional locations of static files
STATICFILES_DIRS = (
    # Put strings here, like "/home/html/static" or "C:/www/django/static".
    # Always use forward slashes, even on Windows.
    # Don't forget to use absolute paths, not relative paths.
)

# List of finder classes that know how to find static files in
# various locations.
STATICFILES_FINDERS = (
    'django.contrib.staticfiles.finders.FileSystemFinder',
    'django.contrib.staticfiles.finders.AppDirectoriesFinder',
)

# Make this unique, and don't share it with anybody.
SECRET_KEY = '6e-b#&0y4mbwu=)hx7a899p(k+i48(p)@e@^aal8^$pn1xqk$$'

TEMPLATES = [
    {
        'BACKEND': 'django.template.backends.django.DjangoTemplates',
        'DIRS': [
            'tests/templates',
        ],
        'OPTIONS': {
            'debug': DEBUG,
            'context_processors': [
                'django.contrib.auth.context_processors.auth',
                'django.contrib.messages.context_processors.messages',
                'django.core.context_processors.debug',
                'django.core.context_processors.i18n',
                'django.core.context_processors.static',
                'django.core.context_processors.media',
                'django.core.context_processors.request',
                'sekizai.context_processors.sekizai',
            ],
            'loaders': [
                'django.template.loaders.filesystem.Loader',
                'django.template.loaders.app_directories.Loader',
                'django.template.loaders.eggs.Loader',
            ]
        },
    },
]

MIDDLEWARE_CLASSES = (
    'django.middleware.common.CommonMiddleware',
    'django.contrib.sessions.middleware.SessionMiddleware',
    'django.middleware.csrf.CsrfViewMiddleware',
    'django.contrib.auth.middleware.AuthenticationMiddleware',
    'django.contrib.messages.middleware.MessageMiddleware',
    # Uncomment the next line for simple clickjacking protection:
    # 'django.middleware.clickjacking.XFrameOptionsMiddleware',
)

ROOT_URLCONF = 'tests.urls'

# Python dotted path to the WSGI application used by Django's runserver.
WSGI_APPLICATION = 'guac_auth.wsgi.application'

INSTALLED_APPS = (
    'django.contrib.auth',
    'django.contrib.contenttypes',
    'django.contrib.sessions',
    'django.contrib.sites',
    'django.contrib.messages',
    'django.contrib.staticfiles',
    'guac_auth',
    # Uncomment the next line to enable the admin:
    # 'django.contrib.admin',
    # Uncomment the next line to enable admin documentation:
    # 'django.contrib.admindocs',
)

# A sample logging configuration. The only tangible logging
# performed by this configuration is to send an email to
# the site admins on every HTTP 500 error when DEBUG=False.
# See http://docs.djangoproject.com/en/dev/topics/logging for
# more details on how to customize your logging configuration.
LOGGING = {
    'version': 1,
    'disable_existing_loggers': False,
    'filters': {
        'require_debug_false': {
            '()': 'django.utils.log.RequireDebugFalse'
        }
    },
    'handlers': {
        'mail_admins': {
            'level': 'ERROR',
            'filters': ['require_debug_false'],
            'class': 'django.utils.log.AdminEmailHandler'
        }
    },
    'loggers': {
        'django.request': {
            'handlers': ['mail_admins'],
            'level': 'ERROR',
            'propagate': True,
        },
    }
}


guac_auth/tests.py
from django.test import TestCase

from guac_auth.models import GuacamoleUser, GuacamoleConnection
from guac_auth.models import GuacamoleConnectionPermission
from guac_auth.utils import quick_rdp, quick_rdp_destroy


class SimpleTestCase(TestCase):
    """
    Simple test cases that verifies basic model functionality.
    """

    def test_quick_rdp(self):
        """
        Test utility method quick_rdp; will exercise models: User, Connection,
        ConnectionParameter, ConnectionPermission, as well other utility
        methods quick_rdp_conn and quick_guac_user.
        """

        # Get a GuacamoleConnectionPermission back.  We can then inspect the
        # Connection and User
        gcp = quick_rdp(
            guac_username="guacman",
            guac_password="avocado",
            username="Administrator",
            password="passwerd",
            hostname="some.windows.host")

        self.assertEquals(gcp.user.username, "guacman")
        self.assertEquals(gcp.connection.protocol, "rdp")

        hostname_parameter = gcp.connection.parameters.get(
            parameter_name='hostname')

        username_parameter = gcp.connection.parameters.get(
            parameter_name='username')

        password_parameter = gcp.connection.parameters.get(
            parameter_name='password')

        self.assertEquals(
            hostname_parameter.parameter_value,
            'some.windows.host')

        self.assertEquals(
            username_parameter.parameter_value,
            'Administrator')

        self.assertEquals(
            password_parameter.parameter_value,
            'passwerd')

        quick_rdp_destroy(
            guac_username="guacman",
            username="Administrator",
            hostname="some.windows.host")

        # We should be empty now
        self.assertEquals(GuacamoleUser.objects.count(), 0)
        self.assertEquals(GuacamoleConnection.objects.count(), 0)
        self.assertEquals(GuacamoleConnectionPermission.objects.count(), 0)


guac_auth/utils.py
import hashlib
import random

from .models import GuacamoleUser, GuacamoleConnection
from .models import GuacamoleConnectionParameter, GuacamoleConnectionPermission


def quick_rdp_conn(username, password, hostname):
    """
    Make a GuacamoleConnection with the passed parameters; return said
    Connection.
    """

    guac_conn = GuacamoleConnection.objects.create(
        connection_name=hostname + ':' + username,
        protocol='rdp')

    # We don't need to save the GuacConnParams as they will CASCADE on_delete
    GuacamoleConnectionParameter.objects.create(
        connection=guac_conn,
        parameter_name='username',
        parameter_value=username)

    GuacamoleConnectionParameter.objects.create(
        connection=guac_conn,
        parameter_name='password',
        parameter_value=password)

    GuacamoleConnectionParameter.objects.create(
        connection=guac_conn,
        parameter_name='hostname',
        parameter_value=hostname)

    return guac_conn


def quick_guac_user(username, password):
    """
    Make a GuacamoleUser with the passed parameters; return said GuacamoleUser.
    Hash password, and salt with random data.
    """

    # MMmm, salt.  Must be 32 bytes.  We must hash with the uppercase,
    # hexadecimal, string representation of the binary.
    salt = bytearray(random.getrandbits(8) for _ in xrange(32))
    salt_hex = ''.join('{:02X}'.format(x) for x in salt)

    # "The salt is appended to the password prior to hashing."
    # http://guac-dev.org/doc/gug/jdbc-auth.html#jdbc-auth-schema-users
    password_hash = hashlib.sha256(password + salt_hex).digest()

    return GuacamoleUser.objects.create(
        username=username,
        password_salt=salt,
        password_hash=password_hash)


def quick_rdp(guac_username, guac_password, username, password, hostname):
    """
    Make a GuacamoleUser and a GuacamoleConnection (RDP) and put them together
    in a GuacamoleConnectionPermission.  Return said
    GuacamoleConnectionPermission.
    """

    guac_conn = quick_rdp_conn(
        username=username,
        password=password,
        hostname=hostname)

    guac_user = quick_guac_user(
        username=guac_username,
        password=guac_password)

    return GuacamoleConnectionPermission.objects.create(
        user=guac_user,
        connection=guac_conn)


def quick_rdp_destroy(guac_username, username, hostname, cleanup_user=True,
                      cleanup_connection=True):
    """
    Remove permissions for this guac user to the guac Connection(s) matching
    these parameters.  May be multiple settings on the same Connection, and/or
    mulitple Connections.  By default, clean up users and Connections that are
    no longer referenced in the Permissions table.
    """

    # Will only get one; username is unique
    user = GuacamoleUser.objects.get(username=guac_username)

    # May be many connections, filter on parameters
    connections = GuacamoleConnection.objects.filter(
        parameters__parameter_name="hostname",
        parameters__parameter_value=hostname).filter(
            parameters__parameter_name="username",
            parameters__parameter_value=username)

    # Take first pass to remove any ConnectionPermission
    for connection in connections:
        GuacamoleConnectionPermission.objects.filter(
            user=user,
            connection=connection).delete()

    # Look again in GuacConnectionPermission.  If we (user or connection) are
    # not found, no one else is using the connection or this user only had the
    # one connection.  Clean up.
    if cleanup_user:
        if GuacamoleConnectionPermission.objects.filter(
                user=user).count() == 0:
            user.delete()

    if cleanup_connection:
        for connection in connections:
            if GuacamoleConnectionPermission.objects.filter(
                    connection=connection).count() == 0:
                connection.delete()


guac_auth/fields.py
from django.db import models


class EnumField(models.Field):
    description = "enumerated type"

    def __init__(self, *args, **kwargs):
        self.enum = kwargs['enum']
        del kwargs['enum']
        super(EnumField, self).__init__(*args, **kwargs)

    def db_type(self, connection):
        return self.enum


class GuacamoleConnectionGroupTypeField(EnumField):
    description = 'enumerated type for connection groups'

    def __init__(self, *args, **kwargs):
        self.enum = 'guacamole_connection_group_type'
        kwargs['enum'] = self.enum
        super(GuacamoleConnectionGroupTypeField,
              self).__init__(*args, **kwargs)


class GuacamoleObjectPermissionTypeField(EnumField):
    description = 'enumerated type for object permissions'

    def __init__(self, *args, **kwargs):
        self.enum = 'guacamole_object_permission_type'
        kwargs['enum'] = self.enum
        super(GuacamoleObjectPermissionTypeField,
              self).__init__(*args, **kwargs)


class GuacamoleSystemPermissionTypeField(EnumField):
    description = 'enumerated type for system permissions'

    def __init__(self, *args, **kwargs):
        self.enum = 'guacamole_system_permission_type'
        kwargs['enum'] = self.enum
        super(GuacamoleSystemPermissionTypeField,
              self).__init__(*args, **kwargs)


guac_auth/models.py
from django.db import models
from django.contrib.auth.models import User

from .fields import GuacamoleConnectionGroupTypeField
from .fields import GuacamoleObjectPermissionTypeField
from .fields import GuacamoleSystemPermissionTypeField

connection_group_type = (
    ('ORGANIZATIONAL', 'ORGANIZATIONAL'),
    ('BALANCING', 'BALANCING'))

object_permission_type = (
    ('READ', 'READ'),
    ('UPDATE', 'UPDATE'),
    ('DELETE', 'DELETE'),
    ('ADMINISTER', 'ADMINISTER'))

system_permission_type = (
    ('CREATE_CONNECTION', 'CREATE_CONNECTION'),
    ('CREATE_CONNECTION_GROUP', 'CREATE_CONNECTION_GROUP'),
    ('CREATE_USER', 'CREATE_USER'),
    ('ADMINISTER', 'ADMINISTER'))


class GuacamoleConnectionGroup(models.Model):
    connection_group_id = models.AutoField(primary_key=True)
    owner = models.ForeignKey(User, null=True)
    parent = models.ForeignKey(
        'self',
        related_name='child_groups',
        blank=True,
        null=True)
    connection_group_name = models.CharField(max_length=128)
    type = GuacamoleConnectionGroupTypeField(
        choices=connection_group_type,
        default='ORGANIZATIONAL')
    max_connections = models.IntegerField(blank=True, null=True)
    max_connections_per_user = models.IntegerField(blank=True, null=True)

    class Meta:
        db_table = 'guacamole_connection_group'
        unique_together = (('connection_group_name', 'parent'),)


class GuacamoleConnection(models.Model):
    connection_id = models.AutoField(primary_key=True)
    connection_name = models.CharField(max_length=128)
    parent = models.ForeignKey(
        GuacamoleConnectionGroup,
        related_name='children',
        blank=True,
        null=True)
    protocol = models.CharField(max_length=32)
    max_connections = models.IntegerField(blank=True, null=True)
    max_connections_per_user = models.IntegerField(blank=True, null=True)

    class Meta:
        db_table = 'guacamole_connection'
        unique_together = (('connection_name', 'parent'),)


class GuacamoleUser(models.Model):
    user_id = models.AutoField(primary_key=True)
    owner = models.ForeignKey(
        User,
        related_name='guacamole_user',
        null=True)
    username = models.CharField(unique=True, max_length=128)
    password_hash = models.BinaryField()
    password_salt = models.BinaryField(blank=True, null=True)
    disabled = models.BooleanField(default=False)
    expired = models.BooleanField(default=False)
    access_window_start = models.TimeField(blank=True, null=True)
    access_window_end = models.TimeField(blank=True, null=True)
    valid_from = models.DateField(blank=True, null=True)
    valid_until = models.DateField(blank=True, null=True)
    timezone = models.CharField(max_length=64, blank=True, null=True)

    class Meta:
        db_table = 'guacamole_user'


class GuacamoleConnectionGroupPermission(models.Model):
    user = models.ForeignKey(
        GuacamoleUser,
        related_name='connection_group_permissions')
    connection_group = models.ForeignKey(GuacamoleConnectionGroup)
    permission = GuacamoleObjectPermissionTypeField(
        choices=object_permission_type,
        default='READ')

    class Meta:
        db_table = 'guacamole_connection_group_permission'
        unique_together = (('user', 'connection_group', 'permission'),)


class GuacamoleConnectionHistory(models.Model):
    history_id = models.AutoField(primary_key=True)
    user = models.ForeignKey(
        GuacamoleUser,
        related_name='history')
    connection = models.ForeignKey(
        GuacamoleConnection,
        related_name='history')
    start_date = models.DateTimeField()
    end_date = models.DateTimeField(blank=True, null=True)

    class Meta:
        db_table = 'guacamole_connection_history'


class GuacamoleConnectionParameter(models.Model):
    connection = models.ForeignKey(
        GuacamoleConnection,
        related_name='parameters',
        on_delete=models.CASCADE)
    parameter_name = models.CharField(max_length=128)
    parameter_value = models.CharField(max_length=4096)

    class Meta:
        db_table = 'guacamole_connection_parameter'
        unique_together = (('connection', 'parameter_name'),)


class GuacamoleConnectionPermission(models.Model):
    user = models.ForeignKey(
        GuacamoleUser,
        related_name='connection_permissions',
        on_delete=models.CASCADE)
    connection = models.ForeignKey(
        GuacamoleConnection,
        related_name='connection_permissions',
        on_delete=models.CASCADE)
    permission = GuacamoleObjectPermissionTypeField(
        choices=object_permission_type,
        default='READ')

    class Meta:
        db_table = 'guacamole_connection_permission'
        unique_together = (('user', 'connection', 'permission'),)


class GuacamoleSystemPermission(models.Model):
    user = models.ForeignKey(
        GuacamoleUser,
        related_name='system_permissions')
    permission = GuacamoleSystemPermissionTypeField(
        choices=system_permission_type,
        default='CREATE_CONNECTION')

    class Meta:
        db_table = 'guacamole_system_permission'
        unique_together = (('user', 'permission'),)


class GuacamoleUserPermission(models.Model):
    user = models.ForeignKey(
        GuacamoleUser,
        related_name='owned_perms')
    affected_user = models.ForeignKey(
        GuacamoleUser,
        related_name='affected_perms')
    permission = GuacamoleObjectPermissionTypeField(
        choices=object_permission_type,
        default='READ')

    class Meta:
        db_table = 'guacamole_user_permission'
        unique_together = (('user', 'affected_user', 'permission'),)


guac_auth/migrations/0002_auto_20160128_1752.py
# -*- coding: utf-8 -*-
from __future__ import unicode_literals

from django.db import migrations, models


class Migration(migrations.Migration):

    dependencies = [
        ('guac_auth', '0001_initial'),
    ]

    operations = [
        migrations.RemoveField(
            model_name='guacamoleconnectiongroup',
            name='type',
        ),
        migrations.AlterUniqueTogether(
            name='guacamoleconnectiongrouppermission',
            unique_together=set([]),
        ),
        migrations.AlterUniqueTogether(
            name='guacamoleconnectionpermission',
            unique_together=set([]),
        ),
        migrations.AlterUniqueTogether(
            name='guacamolesystempermission',
            unique_together=set([]),
        ),
        migrations.AlterUniqueTogether(
            name='guacamoleuserpermission',
            unique_together=set([]),
        ),
        migrations.RemoveField(
            model_name='guacamoleconnectiongrouppermission',
            name='permission',
        ),
        migrations.RemoveField(
            model_name='guacamoleconnectionpermission',
            name='permission',
        ),
        migrations.RemoveField(
            model_name='guacamolesystempermission',
            name='permission',
        ),
        migrations.RemoveField(
            model_name='guacamoleuserpermission',
            name='permission',
        ),
    ]


guac_auth/migrations/0003_auto_20160128_1757.py
# -*- coding: utf-8 -*-
from __future__ import unicode_literals

from django.db import migrations, models, connection
import guac_auth.fields


class Migration(migrations.Migration):

    def enums_forward(apps, schema_editor):
        # SQLite (tests) does not have ENUMs, so don't run then during tests.
        # Required for Postgres however.
        if schema_editor.connection.vendor == 'sqlite':
            return
        cursor = connection.cursor()

        cursor.execute("CREATE TYPE guacamole_connection_group_type AS ENUM( 'ORGANIZATIONAL', 'BALANCING');")
        cursor.execute("CREATE TYPE guacamole_object_permission_type AS ENUM( 'READ', 'UPDATE', 'DELETE', 'ADMINISTER');")
        cursor.execute("CREATE TYPE guacamole_system_permission_type AS ENUM( 'CREATE_CONNECTION', 'CREATE_CONNECTION_GROUP', 'CREATE_USER', 'ADMINISTER');")

    def enums_reverse(apps, schema_editor):
        # SQLite (tests) does not have ENUMs, so don't run then during tests.
        # Required for Postgres however.
        if schema_editor.connection.vendor == 'sqlite':
            return
        cursor = connection.cursor()

        cursor.execute("DROP TYPE guacamole_connection_group_type")
        cursor.execute("DROP TYPE guacamole_object_permission_type")
        cursor.execute("DROP TYPE guacamole_system_permission_type")

    dependencies = [
        ('guac_auth', '0002_auto_20160128_1752'),
    ]

    operations = [
        migrations.RunPython(enums_forward, enums_reverse),
        migrations.AddField(
            model_name='guacamoleconnectiongroup',
            name='type',
            field=guac_auth.fields.GuacamoleConnectionGroupTypeField(default=b'ORGANIZATIONAL', choices=[(b'ORGANIZATIONAL', b'ORGANIZATIONAL'), (b'BALANCING', b'BALANCING')]),
        ),
        migrations.AddField(
            model_name='guacamoleconnectiongrouppermission',
            name='permission',
            field=guac_auth.fields.GuacamoleObjectPermissionTypeField(default=b'READ', choices=[(b'READ', b'READ'), (b'UPDATE', b'UPDATE'), (b'DELETE', b'DELETE'), (b'ADMINISTER', b'ADMINISTER')]),
        ),
        migrations.AddField(
            model_name='guacamoleconnectionpermission',
            name='permission',
            field=guac_auth.fields.GuacamoleObjectPermissionTypeField(default=b'READ', choices=[(b'READ', b'READ'), (b'UPDATE', b'UPDATE'), (b'DELETE', b'DELETE'), (b'ADMINISTER', b'ADMINISTER')]),
        ),
        migrations.AddField(
            model_name='guacamolesystempermission',
            name='permission',
            field=guac_auth.fields.GuacamoleSystemPermissionTypeField(default=b'CREATE_CONNECTION', choices=[(b'CREATE_CONNECTION', b'CREATE_CONNECTION'), (b'CREATE_CONNECTION_GROUP', b'CREATE_CONNECTION_GROUP'), (b'CREATE_USER', b'CREATE_USER'), (b'ADMINISTER', b'ADMINISTER')]),
        ),
        migrations.AddField(
            model_name='guacamoleuserpermission',
            name='permission',
            field=guac_auth.fields.GuacamoleObjectPermissionTypeField(default=b'READ', choices=[(b'READ', b'READ'), (b'UPDATE', b'UPDATE'), (b'DELETE', b'DELETE'), (b'ADMINISTER', b'ADMINISTER')]),
        ),
        migrations.AlterUniqueTogether(
            name='guacamoleconnectiongrouppermission',
            unique_together=set([('user', 'connection_group', 'permission')]),
        ),
        migrations.AlterUniqueTogether(
            name='guacamoleconnectionpermission',
            unique_together=set([('user', 'connection', 'permission')]),
        ),
        migrations.AlterUniqueTogether(
            name='guacamolesystempermission',
            unique_together=set([('user', 'permission')]),
        ),
        migrations.AlterUniqueTogether(
            name='guacamoleuserpermission',
            unique_together=set([('user', 'affected_user', 'permission')]),
        ),
    ]


guac_auth/migrations/0001_initial.py
# -*- coding: utf-8 -*-
from __future__ import unicode_literals

from django.db import migrations, models
from django.conf import settings


class Migration(migrations.Migration):

    dependencies = [
        migrations.swappable_dependency(settings.AUTH_USER_MODEL),
    ]

    operations = [
        migrations.CreateModel(
            name='GuacamoleConnection',
            fields=[
                ('connection_id', models.AutoField(serialize=False, primary_key=True)),
                ('connection_name', models.CharField(max_length=128)),
                ('protocol', models.CharField(max_length=32)),
                ('max_connections', models.IntegerField(null=True, blank=True)),
                ('max_connections_per_user', models.IntegerField(null=True, blank=True)),
            ],
            options={
                'db_table': 'guacamole_connection',
            },
        ),
        migrations.CreateModel(
            name='GuacamoleConnectionGroup',
            fields=[
                ('connection_group_id', models.AutoField(serialize=False, primary_key=True)),
                ('connection_group_name', models.CharField(max_length=128)),
                ('type', models.CharField(default=b'ORGANIZATIONAL', max_length=128, choices=[(b'ORGANIZATIONAL', b'ORGANIZATIONAL'), (b'BALANCING', b'BALANCING')])),
                ('max_connections', models.IntegerField(null=True, blank=True)),
                ('max_connections_per_user', models.IntegerField(null=True, blank=True)),
                ('owner', models.ForeignKey(to=settings.AUTH_USER_MODEL, null=True)),
                ('parent', models.ForeignKey(related_name='child_groups', blank=True, to='guac_auth.GuacamoleConnectionGroup', null=True)),
            ],
            options={
                'db_table': 'guacamole_connection_group',
            },
        ),
        migrations.CreateModel(
            name='GuacamoleConnectionGroupPermission',
            fields=[
                ('id', models.AutoField(verbose_name='ID', serialize=False, auto_created=True, primary_key=True)),
                ('permission', models.CharField(default=b'READ', max_length=128, choices=[(b'READ', b'READ'), (b'UPDATE', b'UPDATE'), (b'DELETE', b'DELETE'), (b'ADMINISTER', b'ADMINISTER')])),
                ('connection_group', models.ForeignKey(to='guac_auth.GuacamoleConnectionGroup')),
            ],
            options={
                'db_table': 'guacamole_connection_group_permission',
            },
        ),
        migrations.CreateModel(
            name='GuacamoleConnectionHistory',
            fields=[
                ('history_id', models.AutoField(serialize=False, primary_key=True)),
                ('start_date', models.DateTimeField()),
                ('end_date', models.DateTimeField(null=True, blank=True)),
                ('connection', models.ForeignKey(related_name='history', to='guac_auth.GuacamoleConnection')),
            ],
            options={
                'db_table': 'guacamole_connection_history',
            },
        ),
        migrations.CreateModel(
            name='GuacamoleConnectionParameter',
            fields=[
                ('id', models.AutoField(verbose_name='ID', serialize=False, auto_created=True, primary_key=True)),
                ('parameter_name', models.CharField(max_length=128)),
                ('parameter_value', models.CharField(max_length=4096)),
                ('connection', models.ForeignKey(related_name='parameters', to='guac_auth.GuacamoleConnection')),
            ],
            options={
                'db_table': 'guacamole_connection_parameter',
            },
        ),
        migrations.CreateModel(
            name='GuacamoleConnectionPermission',
            fields=[
                ('id', models.AutoField(verbose_name='ID', serialize=False, auto_created=True, primary_key=True)),
                ('permission', models.CharField(default=b'READ', max_length=128, choices=[(b'READ', b'READ'), (b'UPDATE', b'UPDATE'), (b'DELETE', b'DELETE'), (b'ADMINISTER', b'ADMINISTER')])),
                ('connection', models.ForeignKey(related_name='connection_permissions', to='guac_auth.GuacamoleConnection')),
            ],
            options={
                'db_table': 'guacamole_connection_permission',
            },
        ),
        migrations.CreateModel(
            name='GuacamoleSystemPermission',
            fields=[
                ('id', models.AutoField(verbose_name='ID', serialize=False, auto_created=True, primary_key=True)),
                ('permission', models.CharField(max_length=128, choices=[(b'CREATE_CONNECTION', b'CREATE_CONNECTION'), (b'CREATE_CONNECTION_GROUP', b'CREATE_CONNECTION_GROUP'), (b'CREATE_USER', b'CREATE_USER'), (b'ADMINISTER', b'ADMINISTER')])),
            ],
            options={
                'db_table': 'guacamole_system_permission',
            },
        ),
        migrations.CreateModel(
            name='GuacamoleUser',
            fields=[
                ('user_id', models.AutoField(serialize=False, primary_key=True)),
                ('username', models.CharField(unique=True, max_length=128)),
                ('password_hash', models.BinaryField()),
                ('password_salt', models.BinaryField(null=True, blank=True)),
                ('disabled', models.BooleanField(default=False)),
                ('expired', models.BooleanField(default=False)),
                ('access_window_start', models.TimeField(null=True, blank=True)),
                ('access_window_end', models.TimeField(null=True, blank=True)),
                ('valid_from', models.DateField(null=True, blank=True)),
                ('valid_until', models.DateField(null=True, blank=True)),
                ('timezone', models.CharField(max_length=64, null=True, blank=True)),
                ('owner', models.ForeignKey(related_name='guacamole_user', to=settings.AUTH_USER_MODEL, null=True)),
            ],
            options={
                'db_table': 'guacamole_user',
            },
        ),
        migrations.CreateModel(
            name='GuacamoleUserPermission',
            fields=[
                ('id', models.AutoField(verbose_name='ID', serialize=False, auto_created=True, primary_key=True)),
                ('permission', models.CharField(default=b'READ', max_length=128, choices=[(b'READ', b'READ'), (b'UPDATE', b'UPDATE'), (b'DELETE', b'DELETE'), (b'ADMINISTER', b'ADMINISTER')])),
                ('affected_user', models.ForeignKey(related_name='affected_perms', to='guac_auth.GuacamoleUser')),
                ('user', models.ForeignKey(related_name='owned_perms', to='guac_auth.GuacamoleUser')),
            ],
            options={
                'db_table': 'guacamole_user_permission',
            },
        ),
        migrations.AddField(
            model_name='guacamolesystempermission',
            name='user',
            field=models.ForeignKey(related_name='system_permissions', to='guac_auth.GuacamoleUser'),
        ),
        migrations.AddField(
            model_name='guacamoleconnectionpermission',
            name='user',
            field=models.ForeignKey(related_name='connection_permissions', to='guac_auth.GuacamoleUser'),
        ),
        migrations.AddField(
            model_name='guacamoleconnectionhistory',
            name='user',
            field=models.ForeignKey(related_name='history', to='guac_auth.GuacamoleUser'),
        ),
        migrations.AddField(
            model_name='guacamoleconnectiongrouppermission',
            name='user',
            field=models.ForeignKey(related_name='connection_group_permissions', to='guac_auth.GuacamoleUser'),
        ),
        migrations.AddField(
            model_name='guacamoleconnection',
            name='parent',
            field=models.ForeignKey(related_name='children', blank=True, to='guac_auth.GuacamoleConnectionGroup', null=True),
        ),
        migrations.AlterUniqueTogether(
            name='guacamoleuserpermission',
            unique_together=set([('user', 'affected_user', 'permission')]),
        ),
        migrations.AlterUniqueTogether(
            name='guacamolesystempermission',
            unique_together=set([('user', 'permission')]),
        ),
        migrations.AlterUniqueTogether(
            name='guacamoleconnectionpermission',
            unique_together=set([('user', 'connection', 'permission')]),
        ),
        migrations.AlterUniqueTogether(
            name='guacamoleconnectionparameter',
            unique_together=set([('connection', 'parameter_name')]),
        ),
        migrations.AlterUniqueTogether(
            name='guacamoleconnectiongrouppermission',
            unique_together=set([('user', 'connection_group', 'permission')]),
        ),
        migrations.AlterUniqueTogether(
            name='guacamoleconnectiongroup',
            unique_together=set([('connection_group_name', 'parent')]),
        ),
        migrations.AlterUniqueTogether(
            name='guacamoleconnection',
            unique_together=set([('connection_name', 'parent')]),
        ),
    ]
